### PR TITLE
[c++] Allow RayConfig to have items other then integer

### PR DESCRIPTION
**Pull request by guoyuhong**
## What do these changes do?
1. Current ray config only has integer items. Some other developers wonder whether we can use other types in ray config and they have to work around to use them. This PR allows other types to be used in ray config.
2. For `id_def.h` and `ray_config_def.h`, some compiler will raise some warning message showed below. 
```
In file included from XXX.cc:12:0:
ray/src/ray/id.h:128:27: warning: extra ‘;’ [-Wpedantic]
 DEFINE_UNIQUE_ID(UniqueID);
                           ^
```

## Related issue number
N/A

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
## Test
Since there is no other types used in ray config now, I test the case of using `double`, `bool`, `std::string` by myself. The code works fine.

### CODE REVIEW

File: src/ray/ray_config.h [hunk 1]
```diff
@@ -35,9 +36,13 @@ class RayConfig {
 /// -----------Include ray_config_def.h to set config items.-------------------
 /// A helper macro that helps to set a value to a config item.
 #define RAY_CONFIG(type, name, default_value) \
-  if (pair.first == #name) { name##_ = pair.second; continue; }
+  if (pair.first == #name) {                  \
+    std::istringstream stream(pair.second);   \
+    stream >> name##_;                        \
+    continue;                                 \
+  }
```
Comment (jovany-wang) [inline src/ray/ray_config.h]: Another approach to reserve the `;` for every `RAY_CONFIG ()` statement is to add a `RAY_IGNORE()` at the end of this macro like:
```C++
#define RAY_CONFIG(type, name, default_value) \
 if (pair.first == #name) {                  \
    std::istringstream stream(pair.second);   \
    stream >> name##_;                        \
    continue;                                 \
  }                                           \
  RAY_IGNORE(nothing)
```

But I don't know if this way is good enough.

**Review by jovany-wang — COMMENTED** [review]
Nice change!

File: src/ray/ray_config.h (hunk 1 above)
Comment (guoyuhong) [reply src/ray/ray_config.h]: That way looks a bit wired. We can wait for other developers' comments.

Comment (jovany-wang) [reply src/ray/ray_config.h]: I don't think this is wired since it's similar to `RAY_LOG()`.

Comment (robertnishihara) [reply src/ray/ray_config.h]: I have a slight preference for requiring a `;` after the macro also. If you add `RAY_IGNORE` at the end, then I'd suggest including a comment to explain its purpose.

Comment (guoyuhong) [reply src/ray/ray_config.h]: OK, will do.

Comment (raulchen) [reply src/ray/ray_config.h]: Adding `RAY_IGNORE` looks fine to me as well.

Comment (guoyuhong) [reply src/ray/ray_config.h]: `RAY_IGNORE_EXPR(0)` can only be used in functions. It is not allowed to use `RAY_IGNORE_EXPR(0)` when a class is defined. I used another way to may the `;` required.

File: src/ray/id.h [hunk 2]
```diff
@@ -51,7 +51,7 @@ std::ostream &operator<<(std::ostream &os, const UniqueID &id);
                                                                                         \
    private:                                                                             \
     type(const std::string &binary) { std::memcpy(id_, binary.data(), kUniqueIDSize); } \
-  };
+  } /*Requires a ; at the end where the macro is used.*/
```
Comment (raulchen) [inline src/ray/id.h]: This way seems more weird. I slightly prefer to just remove the semicolons.

Comment (guoyuhong) [reply src/ray/id.h]: Yes. Since the Macro is only used in `id_def.h` and `ray_config_def.h`, it won't break the traditions in most c++ files.

**Review by raulchen — APPROVED** [review]
LGTM

**Pull request merged**